File: src/main/java/com/server/computerscience/chatbot/service/implement/FileConvertService.java
package com.server.computerscience.chatbot.service.implement;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.List;

import org.springframework.core.io.ByteArrayResource;
import org.springframework.stereotype.Service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.server.computerscience.chatbot.dto.request.ChatGptRequestFileUploadDto;

@Service
public class FileConvertService {
	private final ObjectMapper objectMapper = new ObjectMapper();
	private String jsonFileName = "uploadData";

	public File convertToFile(List<ChatGptRequestFileUploadDto> dataForFile) {
		File jsonlFile = new File("uploadData.jsonl");

		try (BufferedWriter writer = new BufferedWriter(new FileWriter(jsonlFile))) {
			for (ChatGptRequestFileUploadDto dto : dataForFile) {
				// 각 객체를 JSON 문자열로 변환
				String jsonLine = objectMapper.writeValueAsString(dto);
				// 변환된 JSON을 한 줄로 작성
				writer.write(jsonLine);
				writer.newLine();
			}
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}

		return jsonlFile;
	}

	public ByteArrayResource convertToByteArrayResource(List<ChatGptRequestFileUploadDto> dataForFile) {
		try {
			// JSON 데이터 작성
			StringBuilder jsonContent = new StringBuilder();
			ObjectMapper objectMapper = new ObjectMapper();
			for (ChatGptRequestFileUploadDto dto : dataForFile) {
				String jsonLine = objectMapper.writeValueAsString(dto);
				jsonContent.append(jsonLine).append("\n");
			}

			// JSON 데이터를 ByteArrayResource로 변환
			byte[] byteArray = jsonContent.toString().getBytes();
			return new ByteArrayResource(byteArray) {
				@Override
				public String getFilename() {
					return jsonFileName;  // 파일 이름 지정
				}
			};
		} catch (IOException e) {
			e.printStackTrace();
			return null;
		}
	}
}

File: src/main/java/com/server/computerscience/chatbot/service/implement/FileConvertService.kt
package com.server.computerscience.chatbot.service.implement

import com.fasterxml.jackson.databind.ObjectMapper
import com.server.computerscience.chatbot.dto.request.ChatGptRequestFileUploadDto
import org.springframework.core.io.ByteArrayResource
import org.springframework.stereotype.Service
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.OutputStreamWriter

@Service
class FileConvertService {
	private val objectMapper = ObjectMapper()

	fun dataToChatGptJson(dataForFile: List<ChatGptRequestFileUploadDto?>): ByteArrayResource {
		try {
			ByteArrayOutputStream().use { outputStream ->
				OutputStreamWriter(outputStream).use { writer ->
					for (dto in dataForFile) {
						val jsonLine = objectMapper.writeValueAsString(dto)
						writer.write(jsonLine)
						writer.write("\n") // 줄 바꿈 추가
					}
					writer.flush() // OutputStreamWriter 버퍼 비우기
					return object : ByteArrayResource(outputStream.toByteArray()) {
						override fun getFilename(): String = "uploadData.jsonl"
					}
				}
			}
		} catch (e: IOException) {
			e.printStackTrace()
			return ByteArrayResource(ByteArray(0))
		}
	}
}
